### PR TITLE
Add option documentation

### PR DESCRIPTION
**Pull request by jebrando**
Added Option documentation and change all samples to use OPTION_TRUSTED_CERT define

### CODE REVIEW

File: doc/Iothub_sdk_options.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+| `"TrustedCerts"`   | OPTION_TRUSTED_CERT        | const char*        | Azure Server certificate used to validate TLS connection to iothub|
+
+<a name="transport_option"></a>
+
+### Standard Transport
+
+| Option Name            | Option Define             | Value Type         | Description                  |
+|------------------------|---------------------------|--------------------|-------------------------------|
+| `"logtrace"`           | OPTION_LOG_TRACE          | bool* value         | Turn on and off log trace       |
+| `"sas_token_lifetime"` | OPTION_SAS_TOKEN_LIFETIME | `size_t`* value      | Length of time in seconds used for lifetime of sas token. |
```
Comment (zolvarga) [inline doc/Iothub_sdk_options.md]: Beautify pls

File: doc/Iothub_sdk_options.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+| `"keepalive"`          | OPTION_KEEP_ALIVE         | int*               | Length of time to send keep Alives to service
+
+### AMQP Transport
+
+| Option Name            | Option Define                 | Value Type         | Description                  |
+|------------------------|-------------------------------|--------------------|-------------------------------|
+| `"cbs_request_timeout"`| OPTION_CBS_REQUEST_TIMEOUT    | `size_t`* value    | Length of time for CBS token to expire
+| `"sas_token_refresh_time"`| OPTION_SAS_TOKEN_REFRESH_TIME | `size_t`* value    | Length of time to refresh the sas token
+| `"event_send_timeout_secs"`| OPTION_EVENT_SEND_TIMEOUT_SECS | `size_t`* value    | 
+| `"c2d_keep_alive_freq_secs"` | OPTION_C2D_KEEP_ALIVE_FREQ_SECS | `size_t`* value   | Informs service of maximum period the client waits for keep-alive message |
```
Comment (zolvarga) [inline doc/Iothub_sdk_options.md]: Beautify pls

File: doc/Iothub_sdk_options.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+- [Setting an Options](#set_option)
+- [IoThub Client Options](#IotHub_option)
+- [Transport Options](#transport_option)
+
+<a name="set_option"></a>
+
+## Setting an Option
+
+Setting and option in the c-sdk is dependant on which api set you are using.  If using the convience layer you would set the options in the api as follows:
```
Comment (yzhong94) [inline doc/Iothub_sdk_options.md]: Setting and option in the C-SDK is dependent on the API set you use.
If you are using the convenience layer you would set the options in the API as follows:

File: doc/Iothub_sdk_options.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+## Available Option
+
+<a name="IotHub_option"></a>
+
+### IoThub_Client
+
+| Option Name        | Option Define              | Value Type         | Description                  |
+|--------------------|----------------------------|--------------------|-------------------------------|
+| `"messageTimeout"` | OPTION_MESSAGE_TIMEOUT     | tickcounter_ms_t*  | Timeout used for message      |
+| `"product_info"`   | OPTION_PRODUCT_INFO        | const char*        | Product information sent to the IoThub service on connection
```
Comment (yzhong94) [inline doc/Iothub_sdk_options.md]: Can you add more details for the description?  Assuming someone has no idea what any of these features mean?

File: doc/Iothub_sdk_options.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+| `"x509privatekey"`     | OPTION_X509_PRIVATE_KEY   | const char*        | Sets the private key for the x509 certificate
+| `"x509EccCertificate"` | OPTION_X509_ECC_CERT      | const char*        | Sets the ECC x509 certificate used for connection authentication
+| `"x509EccAliasKey"`    | OPTION_X509_ECC_KEY       | const char*        | Sets the private key for the ECC x509 certificate
+| `"proxy_data"`         | OPTION_HTTP_PROXY         | [HTTP_PROXY_OPTIONS*][http-proxy-object]| Http proxy data object used for proxy connection to IoTHub
+
+### MQTT Transport
+
+| Option Name            | Option Define             | Value Type         | Description                  |
+|------------------------|---------------------------|--------------------|-------------------------------|
+| `"keepalive"`          | OPTION_KEEP_ALIVE         | int*               | Length of time to send keep Alives to service
```
Comment (yzhong94) [inline doc/Iothub_sdk_options.md]: What is keep alive?  Are alives sent as d2c message?

File: serializer/samples/simplesample_mqtt/simplesample_mqtt.c [hunk 6]
```diff
@@ -174,7 +175,7 @@ void simplesample_mqtt_run(void)
             {
 #ifdef SET_TRUSTED_CERT_IN_SAMPLES
                 // For mbed add the certificate information
-                if (IoTHubClient_LL_SetOption(iotHubClientHandle, "TrustedCerts", certificates) != IOTHUB_CLIENT_OK)
+                if (IoTHubClient_LL_SetOption(iotHubClientHandle, OPTION_TRUSTED_CERT, certificates) != IOTHUB_CLIENT_OK)
```
Comment (yzhong94) [inline serializer/samples/simplesample_mqtt/simplesample_mqtt.c]: Cool!  For the blog, I can say: Device SDK samples include the option for trusted cert as an example?

File: doc/Iothub_sdk_options.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
+# IoTHub C SDK Options
+
+This document describes the options available in the c sdk and how to use the.
```
Comment (yzhong94) [inline doc/Iothub_sdk_options.md]: Typo

File: doc/Iothub_sdk_options.md (hunk 3 above)
Comment (yzhong94) [reply doc/Iothub_sdk_options.md]: What if the user is not using convenience layer?

Comment (jebrando) [reply doc/Iothub_sdk_options.md]: I'll word that a little better.

---
In reply to: [140867018](https://github.com/Azure/azure-iot-sdk-c/pull/241#discussion_r140867018) [](ancestors = 140867018)

File: serializer/samples/simplesample_mqtt/simplesample_mqtt.c (hunk 6 above)
Comment (jebrando) [reply serializer/samples/simplesample_mqtt/simplesample_mqtt.c]: yep, if you would like to see more examples let me know.

---
In reply to: [140866383](https://github.com/Azure/azure-iot-sdk-c/pull/241#discussion_r140866383) [](ancestors = 140866383)

Comment (yzhong94) [reply serializer/samples/simplesample_mqtt/simplesample_mqtt.c]: Good for now!

File: doc/Iothub_sdk_options.md (hunk 5 above)
Comment (jebrando) [reply doc/Iothub_sdk_options.md]: Keep Alive are for MQTT pings

---
In reply to: [140866080](https://github.com/Azure/azure-iot-sdk-c/pull/241#discussion_r140866080) [](ancestors = 140866080)

File: doc/Iothub_sdk_options.md (hunk 4 above)
Comment (jebrando) [reply doc/Iothub_sdk_options.md]: I was hoping you would tell me what this means.

---
In reply to: [140865655](https://github.com/Azure/azure-iot-sdk-c/pull/241#discussion_r140865655) [](ancestors = 140865655)

Comment (yzhong94) [reply doc/Iothub_sdk_options.md]: Haha :) @tameraw

File: doc/Iothub_sdk_options.md [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+- [Setting an Options](#set_option)
+- [IoThub Client Options](#IotHub_option)
+- [Transport Options](#transport_option)
+
+<a name="set_option"></a>
+
+## Setting an Option
+
+Setting and option in the c-sdk is dependant on which api set you are using:
```
Comment (yzhong94) [inline doc/Iothub_sdk_options.md]: Setting an option

File: doc/Iothub_sdk_options.md [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+IoTHubClient_SetOption(sdk_handle, OPTION_LOG_TRACE, &trace_on);
+
+HTTP_PROXY_OPTIONS http_proxy;
+memset(&http_proxy, 0, sizeof(HTTP_PROXY_OPTIONS));
+http_proxy.host_address = PROXY_ADDRESS;
+http_proxy.port = PROXY_PORT;
+DPS_LL_SetOption(handle, OPTION_HTTP_PROXY, &http_proxy);
+```
+
+## Available Option
```
Comment (tameraw) [inline doc/Iothub_sdk_options.md]: >Option [](start = 13, length = 6)

Options

File: doc/Iothub_sdk_options.md [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+<a name="transport_option"></a>
+
+### Standard Transport
+
+| Option Name            | Option Define             | Value Type         | Description
+|------------------------|---------------------------|--------------------|-------------------------------
+| `"logtrace"`           | OPTION_LOG_TRACE          | bool* value        | Turn on and off log tracing for the transport
+| `"sas_token_lifetime"` | OPTION_SAS_TOKEN_LIFETIME | `size_t`* value    | Length of time in seconds used for lifetime of sas token.
+| `"x509certificate"`    | OPTION_X509_CERT          | const char*        | Sets the x509 certificate used for connection authentication
```
Comment (tameraw) [inline doc/Iothub_sdk_options.md]: >the x509 [](start = 81, length = 8)

Can we clarify that this is RSA?

File: doc/Iothub_sdk_options.md [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+| `"x509privatekey"`     | OPTION_X509_PRIVATE_KEY   | const char*        | Sets the private key for the x509 certificate
+| `"x509EccCertificate"` | OPTION_X509_ECC_CERT      | const char*        | Sets the ECC x509 certificate used for connection authentication
+| `"x509EccAliasKey"`    | OPTION_X509_ECC_KEY       | const char*        | Sets the private key for the ECC x509 certificate
+| `"proxy_data"`         | OPTION_HTTP_PROXY         | [HTTP_PROXY_OPTIONS*][http-proxy-object]| Http proxy data object used for proxy connection to IoTHub
```
Comment (tameraw) [inline doc/Iothub_sdk_options.md]: >Http proxy data object used for proxy connection to IoTHub [](start = 97, length = 58)

Can we break down the members of that object?

File: doc/Iothub_sdk_options.md [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+
+### MQTT Transport
+
+| Option Name            | Option Define             | Value Type         | Description
+|------------------------|---------------------------|--------------------|-------------------------------
+| `"keepalive"`          | OPTION_KEEP_ALIVE         | int*               | Length of time to send `Keep Alives` to service for D2C Messages
```
Comment (tameraw) [inline doc/Iothub_sdk_options.md]: >to service for D2C Messages [](start = 113, length = 27)

I thought this was for setting the C2D keepalives. D2C keepalives are controlled from the service.

File: doc/Iothub_sdk_options.md [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+### AMQP Transport
+
+| Option Name                  | Option Define                   | Value Type        | Description
+|------------------------------|---------------------------------|-------------------|-------------------------------
+| `"cbs_request_timeout"`      | OPTION_CBS_REQUEST_TIMEOUT      | `size_t`* value   | Length of time for CBS token to expire
```
Comment (tameraw) [inline doc/Iothub_sdk_options.md]: >CBS token to expire [](start = 106, length = 19)

I don't think this is for the token expiration. I believe this is for the put request over CBS. Please confirm.

File: doc/Iothub_sdk_options.md [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+| `"sas_token_refresh_time"`   | OPTION_SAS_TOKEN_REFRESH_TIME   | `size_t`* value   | Length of time to refresh the sas token
```
Comment (tameraw) [inline doc/Iothub_sdk_options.md]: >sas token [](start = 117, length = 9)

Please specify unit.

File: doc/Iothub_sdk_options.md (hunk 11 above)
Comment (jebrando) [reply doc/Iothub_sdk_options.md]: I was afraid you'd say that.  It becomes difficult.


---
In reply to: [140919369](https://github.com/Azure/azure-iot-sdk-c/pull/241#discussion_r140919369) [](ancestors = 140919369)

File: doc/Iothub_sdk_options.md (hunk 12 above)
Comment (jebrando) [reply doc/Iothub_sdk_options.md]: This is MQTT, AMQP is controlled by the service.

---
In reply to: [140919574](https://github.com/Azure/azure-iot-sdk-c/pull/241#discussion_r140919574) [](ancestors = 140919574)

**Pull request closed without merging**